src/lib/Scenes/Home/Components/ForYou/index.tsx

```diff
@@ -1,104 +1,126 @@
-    rows.push({ type: "fairs", data: fairsModule })
-    for (let i = 0; i < artworkModules.length; i++) {
-      const artworkModule = artworkModules[i]
-      rows.push({ type: "artwork", data: artworkModule })
-      if ((i + 1) % 2 === 0) {
+    const { forYou } = this.props
+    const rowData = []
+    const artworkModules = forYou.artwork_modules || []
+    const artistModules = forYou.artist_modules && forYou.artist_modules.concat()
+    const fairsModule = forYou.fairs_module
+
+    rowData.push({
+      type: "fairs",
+      data: fairsModule,
+    })
+
+    artworkModules.forEach((artworkModule, index) => {
+      rowData.push({
+        type: "artwork",
+        data: artworkModule,
+      })
+
+      const alternateRow = (index + 1) % 2 === 0
+
+      if (alternateRow) {
         const artistModule = artistModules.shift()
         if (artistModule) {
-          rows.push({ type: "artist", data: artistModule })
+          rowData.push({
+            type: "artist",
+            data: artistModule,
+          })
         }
       }
-    }
+    })
+
     this.setState({
-      dataSource: this.state.dataSource.cloneWithRows(rows),
+      rowData,
     })
   }
 
-  handleRefresh = () => {
-    this.setState({ isRefreshing: true })
-    const stopRefreshing = () => this.setState({ isRefreshing: false })
-
-    Promise.all(
-      this.state.modules.map(module => {
-        return new Promise((resolve, reject) => {
-          module.forceFetch(null, readyState => {
-            if (readyState.error) {
-              reject(readyState.error)
-            } else if (readyState.aborted) {
-              reject()
-            } else if (readyState.done) {
-              resolve()
-            }
-          })
-        })
+  registerRailModule = railModule => {
+    this.railModules = this.railModules.concat([railModule])
+  }
+
+  handleRefresh = async () => {
+    this.setState({
+      isRefreshing: true,
+    })
+
+    try {
+      await Promise.all(this.railModules.map(railModule => railModule.refreshData()))
+
+      this.setState({
+        isRefreshing: false,
+      })
+    } catch (error) {
+      console.error("ForYou/index.tsx", error.message)
+
+      this.setState({
+        error: error.message, // TODO: Display this somehow
       })
-    ).then(stopRefreshing, stopRefreshing)
+    }
+  }
+
+  // TODO: Is this still needed?
+  handleScroll = event => {
+    this.currentScrollOffset = event.nativeEvent.contentOffset.y
   }
 
   render() {
     return (
-      <ListView
-        dataSource={this.state.dataSource}
-        renderScrollComponent={(props: ScrollViewProps) => {
-          const refreshControl = <RefreshControl refreshing={this.state.isRefreshing} onRefresh={this.handleRefresh} />
-          return <ScrollView {...props} automaticallyAdjustContentInsets={false} refreshControl={refreshControl} />
+      <FlatList
+        data={this.state.rowData}
+        renderScrollComponent={props => {
```

Comment:
I'm actually not sure if this prop is available in `FlatList` I see it here (https://facebook.github.io/react-native/docs/listview.html#renderscrollcomponent) but I haven't found any references to it in the docs.

Reply:
If you dig into the component its part of this type:
```
export interface VirtualizedListProperties<ItemT> extends ScrollViewProperties {
```
I assumed it was part of a base class and everything seems to be wired up inside. Maybe undocumented?

Reply:
Got it, yeah they probably didn't document it. 

Definitely, VScode makes it dumb easy to get to the type defs too 👍